### PR TITLE
WIP: Policy: Separate standard and testing policies

### PR DESCRIPTION
**Pull request by jtimon**
EDIT:

This prepares AcceptToMemoryPool to be decoupled from Params() and creates a -policy command-line option to select the standard or test policy independently of the chain mode (ie main, testnet3, regtest).

It's based on #6068.
@petertodd already acked the concept here: https://github.com/bitcoin/bitcoin/pull/5521#issuecomment-67774957

OLD:
Instead of being based on #5071 as initially, it doesn't go as far as that PR. 
A rebased version of #5071 on top of this PR can be found here: https://github.com/jtimon/bitcoin/tree/nodepolicy_5180
Instead of being based on #5071 it is based on #5521 (plus one doubt-commit).

Please let's not delay policy.o even if it's only with `minRelayTxFee`.
Later proposals to expand policy.o become more readable once the first step has been walked.
Even better, later proposed changes to main.o may become proposed changes to policy.o!!
In my opinion the biggest development bottleneck is reviewing changes to main.o.

/EDIT
It's based on #5071 and eliminates Params().RequireStandard() by separating two policies for the 4 networks accessible with Params() as discussed in that PR.
It also includes a couple of "squashme" commits that solve my complains on #5071 in case you @luke-jr want to take them.

### CODE REVIEW

**Comment by jtimon** [comment]
Updated code and description again without IsDust() this time

**Comment by jtimon** [comment]
Rebased after #5521 has been merged for easier review.

**Comment by jtimon** [comment]
Closing in favor of #5652

**Comment by jtimon** [comment]
Updated after #6068 has been updated. Now the addition of the -policy option is left to the end as an optinal "squasheable" commit.

**Comment by jtimon** [comment]
Closing until some version of #6068 is merged.

**Comment by jtimon** [comment]
Updated opening description after rebasing and re-closing.

File: src/init.cpp [hunk 1]
```diff
@@ -820,9 +819,8 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
             return InitError(strprintf(_("Invalid amount for -minrelaytxfee=<amount>: '%s'"), mapArgs["-minrelaytxfee"]));
     }
 
-    fRequireStandard = !GetBoolArg("-acceptnonstdtxn", !Params().RequireStandard());
-    if (Params().RequireStandard() && !fRequireStandard)
-        return InitError(strprintf("acceptnonstdtxn is not currently supported for %s chain", chainparams.NetworkIDString()));
+    if (GetBoolArg("-acceptnonstdtxn", false))
+        return InitError(strprintf("acceptnonstdtxn is deprecated, use policy=test instead.", chainparams.NetworkIDString()));
```
Comment (luke-jr) [inline src/init.cpp]: acceptnonstdtxn affects only one facet of the policy. It shouldn't be deprecated unless you add a way to configure the flag some other way.

Comment (jtimon) [inline src/init.cpp]: The default -acceptnonstdtxn=0 is equivalent to the default -policy=standard.
To use the equivalent of -acceptnonstdtxn=1, you use -policy=test.
This has the advantages of decoupling policy from Params() and that adding new policies (say, -policy=fss_rbf) becomes easier in the future.
Anyway, this needs rebase.

**Comment by jtimon** [comment]
Updated without deprecating anything.
Now -policy overwrites the chain selection but -acceptnonstdtxn overwrites -policy (at least for policies standard and test, other policies may not even have a -acceptnonstdtxn option).
I don't care about keeping -acceptnonstdtxn even if you can just use -policy=test instead of -acceptnonstdtxn=1 (so keeping it is redundant), to me the important thing is decoupling policy from Params() and having a factory to make the creation maintenance of custom policies easier.

**Comment by jtimon** [comment]
I only reopened to show and discuss the updated version. But, again, closing until there's a CPolicy interface class.

**Pull request closed without merging**